--- src/dird/msgchan.c  8 Dec 2006 14:27:10 -0000       1.66
+++ src/dird/msgchan.c  5 Feb 2007 14:15:52 -0000
@@ -316,7 +316,7 @@
       Jmsg1(jcr, M_ABORT, 0, _("Cannot create message thread: %s\n"), be.strerror(status));
    }
    /* Wait for thread to start */
-   while (jcr->SD_msg_chan == 0) {
+   while (jcr->SD_msg_chan == 0 && cr->sd_msg_thread_done == false) {
       bmicrosleep(0, 50);
    }
    Dmsg1(100, "SD msg_thread started. use=%d\n", jcr->use_count());
