Index: src/dird/dird_conf.c
===================================================================
--- src/dird/dird_conf.c	(révision 7525)
+++ src/dird/dird_conf.c	(copie de travail)
@@ -509,6 +509,16 @@
    {NULL,               0}
 };
 
+char *CAT::display(POOLMEM *dst) {
+   Mmsg(dst,"catalog=%s\ndb_name=%s\ndb_driver=%s\ndb_user=%s\n"
+        "db_password=%s\ndb_address=%s\ndb_port=%i\n"
+        "db_socket=%s\n",
+        name(), NPRTB(db_name),
+        NPRTB(db_driver), NPRTB(db_user), NPRTB(db_password),
+        NPRTB(db_address), db_port, NPRTB(db_socket));
+   return dst;
+}
+
 const char *level_to_str(int level)
 {
    int i;
Index: src/dird/dird_conf.h
===================================================================
--- src/dird/dird_conf.h	(révision 7525)
+++ src/dird/dird_conf.h	(copie de travail)
@@ -236,6 +236,7 @@
 
    /* Methods */
    char *name() const;
+   char *display(POOLMEM *dst);       /* Get catalog information */
 };
 
 inline char *CAT::name() const { return hdr.name; }
Index: src/baconfig.h
===================================================================
--- src/baconfig.h	(révision 7525)
+++ src/baconfig.h	(copie de travail)
@@ -79,7 +79,8 @@
 
 /* Allow printing of NULL pointers */
 #define NPRT(x) (x)?(x):_("*None*")
- 
+#define NPRTB(x) (x)?(x):""
+
 #if defined(HAVE_WIN32)
 
 #define WIN32_REPARSE_POINT 1
Index: src/tools/dbcheck.c
===================================================================
--- src/tools/dbcheck.c	(révision 7525)
+++ src/tools/dbcheck.c	(copie de travail)
@@ -100,14 +100,14 @@
 static void do_interactive_mode();
 static bool yes_no(const char *prompt);
 
-
 static void usage()
 {
    fprintf(stderr,
-"Usage: dbcheck [-c config] [-C catalog name] [-d debug_level] <working-directory> <bacula-database> <user> <password> [<dbhost>] [<dbport>]\n"
+"Usage: dbcheck [-c config ] [-B] [-C catalog name] [-d debug_level] <working-directory> <bacula-database> <user> <password> [<dbhost>] [<dbport>]\n"
 "       -b              batch mode\n"
 "       -C              catalog name in the director conf file\n"
 "       -c              Director conf filename\n"
+"       -B              print catalog configuration and exit\n"
 "       -d <nn>         set debug level to <nn>\n"
 "       -dt             print timestamp in debug output\n"
 "       -f              fix inconsistencies\n"
@@ -121,6 +121,7 @@
    int ch;
    const char *user, *password, *db_name, *dbhost;
    int dbport = 0;
+   bool print_catalog=false;
    char *configfile = NULL;
    char *catalogname = NULL;
    char *endptr;
@@ -136,8 +137,12 @@
    memset(&name_list, 0, sizeof(name_list));
 
 
-   while ((ch = getopt(argc, argv, "bc:C:d:fv?")) != -1) {
+   while ((ch = getopt(argc, argv, "bc:C:d:fvB?")) != -1) {
       switch (ch) {
+      case 'B':
+         print_catalog = true;     /* get catalog information from config */
+         break;
+
       case 'b':                    /* batch */
          batch = true;
          break;
@@ -215,6 +220,16 @@
             exit(1);
          }
          set_working_directory(director->working_directory);
+
+         /* Print catalog information and exit (-B) */
+         if (print_catalog) {
+            POOLMEM *buf = get_pool_memory(PM_MESSAGE);
+            printf("%sdb_type=%s\nworking_dir=%s\n", catalog->display(buf),
+                   db_get_type(), working_directory);
+            free_pool_memory(buf);
+            exit(0);
+         }
+
          db_name = catalog->db_name;
          user = catalog->db_user;
          password = catalog->db_password;
