--- Makefile	2008-06-27 17:15:12.000000000 +0200
+++ Makefile.mingw	2008-06-27 17:15:04.000000000 +0200
@@ -22,7 +22,7 @@
 # Targets
 HEADERFILES = tcutil.h tchdb.h tcbdb.h tcadb.h
 LIBRARYFILES = libtokyocabinet.a libtokyocabinet.so.3.6.0 libtokyocabinet.so.3 libtokyocabinet.so
-LIBOBJFILES = tcutil.o tchdb.o tcbdb.o tcadb.o myconf.o
+LIBOBJFILES = tcutil.o tchdb.o tcbdb.o tcadb.o myconf.o compat.o
 COMMANDFILES = tcutest tcumttest tcucodec tchtest tchmttest tchmgr tcbtest tcbmttest tcbmgr tcatest tcamgr
 MAN1FILES = tcutest.1 tcumttest.1 tcucodec.1 tchtest.1 tchmttest.1 tchmgr.1 tcbtest.1 tcbmttest.1 tcbmgr.1 tcatest.1 tcamgr.1
 MAN3FILES = tokyocabinet.3 tcutil.3 tcxstr.3 tclist.3 tcmap.3 tcmdb.3 tcmpool.3 tchdb.3 tcbdb.3 tcadb.3
@@ -45,17 +45,17 @@
 
 # Building configuration
 CC = i586-mingw32msvc-gcc
-CPPFLAGS = -I. -I$(INCLUDEDIR) -L/home/eric/include -L/usr/local/include -DNDEBUG -D_GNU_SOURCE=1 -UNDEBUG \
+CPPFLAGS = -I. -I$(INCLUDEDIR) -I/home/eric/include -L/usr/local/include -DNDEBUG -D_GNU_SOURCE=1 -UNDEBUG \
   -D_TC_PREFIX="\"$(prefix)\"" -D_TC_INCLUDEDIR="\"$(INCLUDEDIR)\"" \
   -D_TC_LIBDIR="\"$(LIBDIR)\"" -D_TC_BINDIR="\"$(BINDIR)\"" -D_TC_LIBEXECDIR="\"$(LIBEXECDIR)\"" \
   -D_TC_APPINC="\"-I$(INCLUDEDIR)\"" -D_TC_APPLIBS="\"-L$(LIBDIR) -ltokyocabinet -lm \""
-CFLAGS = -std=c99 -Wall -fPIC -pedantic -fsigned-char -g
+CFLAGS = --std=c99 -Wall -pedantic -fsigned-char -g -DHAVE_WIN32 
 LDFLAGS = -L. -L$(LIBDIR) -L/home/eric/lib -L/usr/local/lib -static
-LIBS = -lm 
+LIBS = -lm -lz -lpthreadGCE
 LDENV = LD_RUN_PATH=/lib:/usr/lib:$(LIBDIR):$(HOME)/lib:/usr/local/lib:$(LIBDIR):.
 RUNENV = LD_LIBRARY_PATH=.:/lib:/usr/lib:$(LIBDIR):$(HOME)/lib:/usr/local/lib:$(LIBDIR)
 POSTCMD = true
-NO_ECHO = @
+NO_ECHO = 
 
 
 #================================================================
@@ -447,47 +447,47 @@
 	ln -f -s libtokyocabinet.$(LIBVER).$(LIBREV).0.dylib $@
 
 
-tcutest : tcutest.o $(LIBRARYFILES)
+tcutest.exe : tcutest.o $(LIBRARYFILES)
 	$(LDENV) $(CC) $(CFLAGS) -o $@ $< $(LDFLAGS) -ltokyocabinet $(LIBS)
 
 
-tcumttest : tcumttest.o $(LIBRARYFILES)
+tcumttest.exe : tcumttest.o $(LIBRARYFILES)
 	$(LDENV) $(CC) $(CFLAGS) -o $@ $< $(LDFLAGS) -ltokyocabinet $(LIBS)
 
 
-tcucodec : tcucodec.o $(LIBRARYFILES)
+tcucodec.exe : tcucodec.o $(LIBRARYFILES)
 	$(LDENV) $(CC) $(CFLAGS) -o $@ $< $(LDFLAGS) -ltokyocabinet $(LIBS)
 
 
-tchtest : tchtest.o $(LIBRARYFILES)
+tchtest.exe : tchtest.o $(LIBRARYFILES)
 	$(LDENV) $(CC) $(CFLAGS) -o $@ $< $(LDFLAGS) -ltokyocabinet $(LIBS)
 
 
-tchmttest : tchmttest.o $(LIBRARYFILES)
+tchmttest.exe : tchmttest.o $(LIBRARYFILES)
 	$(LDENV) $(CC) $(CFLAGS) -o $@ $< $(LDFLAGS) -ltokyocabinet $(LIBS)
 
 
-tchmgr : tchmgr.o $(LIBRARYFILES)
+tchmgr.exe : tchmgr.o $(LIBRARYFILES)
 	$(LDENV) $(CC) $(CFLAGS) -o $@ $< $(LDFLAGS) -ltokyocabinet $(LIBS)
 
 
-tcbtest : tcbtest.o $(LIBRARYFILES)
+tcbtest.exe : tcbtest.o $(LIBRARYFILES)
 	$(LDENV) $(CC) $(CFLAGS) -o $@ $< $(LDFLAGS) -ltokyocabinet $(LIBS)
 
 
-tcbmttest : tcbmttest.o $(LIBRARYFILES)
+tcbmttest.exe : tcbmttest.o $(LIBRARYFILES)
 	$(LDENV) $(CC) $(CFLAGS) -o $@ $< $(LDFLAGS) -ltokyocabinet $(LIBS)
 
 
-tcbmgr : tcbmgr.o $(LIBRARYFILES)
+tcbmgr.exe : tcbmgr.o $(LIBRARYFILES)
 	$(LDENV) $(CC) $(CFLAGS) -o $@ $< $(LDFLAGS) -ltokyocabinet $(LIBS)
 
 
-tcatest : tcatest.o $(LIBRARYFILES)
+tcatest.exe : tcatest.o $(LIBRARYFILES)
 	$(LDENV) $(CC) $(CFLAGS) -o $@ $< $(LDFLAGS) -ltokyocabinet $(LIBS)
 
 
-tcamgr : tcamgr.o $(LIBRARYFILES)
+tcamgr.exe : tcamgr.o $(LIBRARYFILES)
 	$(LDENV) $(CC) $(CFLAGS) -o $@ $< $(LDFLAGS) -ltokyocabinet $(LIBS)
 
 
