Index: myconf.h
===================================================================
--- myconf.h	(revision 7238)
+++ myconf.h	(working copy)
@@ -206,13 +206,18 @@
 #include <unistd.h>
 #include <sys/types.h>
 #include <sys/stat.h>
-#include <sys/mman.h>
 #include <sys/time.h>
-#include <sys/times.h>
 #include <fcntl.h>
 #include <dirent.h>
+
+#ifdef HAVE_WIN32
+#include "compat.h"
+#else
+#include <sys/mman.h>
+#include <sys/times.h>
 #include <regex.h>
 #include <glob.h>
+#endif
 
 #if TCUSEPTHREAD
 #include <pthread.h>
Index: tcutil.c
===================================================================
--- tcutil.c	(revision 7238)
+++ tcutil.c	(working copy)
@@ -17,7 +17,6 @@
 #include "tcutil.h"
 #include "myconf.h"
 
-
 /*************************************************************************************************
  * basic utilities
  *************************************************************************************************/
@@ -3333,6 +3332,7 @@
 
 /* Lock a file. */
 bool tclock(int fd, bool ex, bool nb){
+#ifndef HAVE_WIN32
   assert(fd >= 0);
   struct flock lock;
   memset(&lock, 0, sizeof(struct flock));
@@ -3344,6 +3344,7 @@
   while(fcntl(fd, nb ? F_SETLK : F_SETLKW, &lock) == -1){
     if(errno != EINTR) return false;
   }
+#endif
   return true;
 }
 
